src/persist-client/src/write.rs

```diff
@@ -175,11 +175,23 @@ where
 
     /// A cached version of the shard-global `upper` frontier.
     ///
-    /// This will always be less or equal to the shard-global `upper`.
+    /// This is the most recent upper discovered by this handle. It is
+    /// potentially more stale than [Self::pubsub_upper] but is lock-free and
+    /// allocation-free. This will always be less or equal to the shard-global
+    /// `upper`.
     pub fn upper(&self) -> &Antichain<T> {
         &self.upper
     }
 
+    /// A less-stale cached version of the shard-global `upper` frontier.
+    ///
+    /// This is the most recently known upper for this shard process-wide, but
+    /// unlike [Self::upper] it requires a mutex and a clone. This will always be
+    /// less or equal to the shard-global `upper`.
+    pub fn pubsub_upper(&self) -> Antichain<T> {
```

Comment:
I'd suggest `shared_upper` or `cached_upper` since `pubsub_upper` feels misleading to me (IIUC it advances even if no pubsub is happening) but don't feel strongly.

Reply:
I didn't love the name I had either, so shared_upper sgtm! Done